### Background
Logistics in Sub-Saharan Africa increases the cost of manufactured goods by up to 320%; while in Europe, it only accounts for up to 90% of the manufacturing cost. Sendy is a business-to-business platform established in 2014, to enable businesses of all types and sizes to transport goods more efficiently across East Africa. The company is headquartered in Kenya with a team of more than 100 staff, focused on building practical solutions for Africa’s dynamic transportation needs, from developing apps and web solutions to providing dedicated support for goods on the move.
#### Problem Statement
Sendy has hired you to help predict the estimated time of delivery of orders, from the point of driver pickup to the point of arrival at the final destination. Build a model that predicts an accurate delivery time, from picking up a package arriving at the final destination. An accurate arrival time prediction will help all business to improve their logistics and communicate the accurate time their time to their customers. You will be required to perform various feature engineering techniques while preparing your data for further analysis.

In [64]:
# Load the necessary libraries

import pandas as pd

In [65]:
# Load the Data set
df = pd.read_csv('https://bit.ly/3deaKEM')


In [66]:
#read the first five records
df.head()

Order No       User Id Vehicle Type  Platform Type  \
0   Order_No_4211   User_Id_633         Bike              3   
1  Order_No_25375  User_Id_2285         Bike              3   
2   Order_No_1899   User_Id_265         Bike              3   
3   Order_No_9336  User_Id_1402         Bike              3   
4  Order_No_27883  User_Id_1737         Bike              1   

  Personal or Business  Placement - Day of Month  \
0             Business                         9   
1             Personal                        12   
2             Business                        30   
3             Business                        15   
4             Personal                        13   

   Placement - Weekday (Mo = 1) Placement - Time  Confirmation - Day of Month  \
0                             5       9:35:46 AM                            9   
1                             5      11:16:16 AM                           12   
2                             2      12:39:25 PM                           30   
3                             5       9:25:34 AM                           15   
4                             1       9:55:18 AM                           13   

   Confirmation - Weekday (Mo = 1)  ... Arrival at Destination - Time  \
0                                5  ...                   10:39:55 AM   
1                                5  ...                   12:17:22 PM   
2                                2  ...                    1:00:38 PM   
3                                5  ...                   10:05:27 AM   
4                                1  ...                   10:25:37 AM   

   Distance (KM)  Temperature Precipitation in millimeters  Pickup Lat  \
0              4         20.4                          NaN   -1.317755   
1             16         26.4                          NaN   -1.351453   
2              3          NaN                          NaN   -1.308284   
3              9         19.2                          NaN   -1.281301   
4              9         15.4                          NaN   -1.266597   

   Pickup Long Destination Lat  Destination Long      Rider Id  \
0    36.830370       -1.300406         36.829741  Rider_Id_432   
1    36.899315       -1.295004         36.814358  Rider_Id_856   
2    36.843419       -1.300921         36.828195  Rider_Id_155   
3    36.832396       -1.257147         36.795063  Rider_Id_855   
4    36.792118       -1.295041         36.809817  Rider_Id_770   

  Time from Pickup to Arrival  
0                         745  
1                        1993  
2                         455  
3                        1341  
4                        1214  

[5 rows x 29 columns]

In [67]:
# View the last five records
df.tail()

Order No       User Id Vehicle Type  Platform Type  \
21196   Order_No_8834  User_Id_2001         Bike              3   
21197  Order_No_22892  User_Id_1796         Bike              3   
21198   Order_No_2831  User_Id_2956         Bike              3   
21199   Order_No_6174  User_Id_2524         Bike              1   
21200   Order_No_9836   User_Id_718         Bike              3   

      Personal or Business  Placement - Day of Month  \
21196             Personal                        20   
21197             Business                        13   
21198             Business                         7   
21199             Personal                         4   
21200             Business                        26   

       Placement - Weekday (Mo = 1) Placement - Time  \
21196                             3       3:54:38 PM   
21197                             6      10:13:34 AM   
21198                             4       5:06:16 PM   
21199                             3       9:31:39 AM   
21200                             2       2:19:47 PM   

       Confirmation - Day of Month  Confirmation - Weekday (Mo = 1)  ...  \
21196                           20                                3  ...   
21197                           13                                6  ...   
21198                            7                                4  ...   
21199                            4                                3  ...   
21200                           26                                2  ...   

      Arrival at Destination - Time  Distance (KM)  Temperature  \
21196                    4:20:17 PM              3         28.6   
21197                   10:46:17 AM              7         26.0   
21198                    6:40:05 PM             20         29.2   
21199                   10:08:15 AM             13         15.0   
21200                    3:17:23 PM             12         30.9   

      Precipitation in millimeters  Pickup Lat  Pickup Long Destination Lat  \
21196                          NaN   -1.258414    36.804800       -1.275285   
21197                          NaN   -1.307143    36.825009       -1.331619   
21198                          NaN   -1.286018    36.897534       -1.258414   
21199                          NaN   -1.250030    36.874167       -1.279210   
21200                          NaN   -1.255189    36.782203       -1.320157   

       Destination Long      Rider Id Time from Pickup to Arrival  
21196         36.802702  Rider_Id_953                           9  
21197         36.847976  Rider_Id_155                         770  
21198         36.804800  Rider_Id_697                        2953  
21199         36.794872  Rider_Id_347                        1380  
21200         36.830887  Rider_Id_177                        2128  

[5 rows x 29 columns]

In [68]:
# View the data types
df.dtypes

Order No                                      object
User Id                                       object
Vehicle Type                                  object
Platform Type                                  int64
Personal or Business                          object
Placement - Day of Month                       int64
Placement - Weekday (Mo = 1)                   int64
Placement - Time                              object
Confirmation - Day of Month                    int64
Confirmation - Weekday (Mo = 1)                int64
Confirmation - Time                           object
Arrival at Pickup - Day of Month               int64
Arrival at Pickup - Weekday (Mo = 1)           int64
Arrival at Pickup - Time                      object
Pickup - Day of Month                          int64
Pickup - Weekday (Mo = 1)                      int64
Pickup - Time                                 object
Arrival at Destination - Day of Month          int64
Arrival at Destination - Weekday (Mo = 1)     

In [69]:
# View the descriptive statistics
df.describe()

Platform Type  Placement - Day of Month  Placement - Weekday (Mo = 1)  \
count   21201.000000              21201.000000                  21201.000000   
mean        2.752182                 15.653696                      3.240083   
std         0.625178                  8.798916                      1.567295   
min         1.000000                  1.000000                      1.000000   
25%         3.000000                  8.000000                      2.000000   
50%         3.000000                 15.000000                      3.000000   
75%         3.000000                 23.000000                      5.000000   
max         4.000000                 31.000000                      7.000000   

       Confirmation - Day of Month  Confirmation - Weekday (Mo = 1)  \
count                 21201.000000                     21201.000000   
mean                     15.653837                         3.240225   
std                       8.798886                         1.567228   
min                       1.000000                         1.000000   
25%                       8.000000                         2.000000   
50%                      15.000000                         3.000000   
75%                      23.000000                         5.000000   
max                      31.000000                         7.000000   

       Arrival at Pickup - Day of Month  Arrival at Pickup - Weekday (Mo = 1)  \
count                      21201.000000                          21201.000000   
mean                          15.653837                              3.240225   
std                            8.798886                              1.567228   
min                            1.000000                              1.000000   
25%                            8.000000                              2.000000   
50%                           15.000000                              3.000000   
75%                           23.000000                              5.000000   
max                           31.000000                              7.000000   

       Pickup - Day of Month  Pickup - Weekday (Mo = 1)  \
count           21201.000000               21201.000000   
mean               15.653837                   3.240225   
std                 8.798886                   1.567228   
min                 1.000000                   1.000000   
25%                 8.000000                   2.000000   
50%                15.000000                   3.000000   
75%                23.000000                   5.000000   
max                31.000000                   7.000000   

       Arrival at Destination - Day of Month  \
count                           21201.000000   
mean                               15.653837   
std                                 8.798886   
min                                 1.000000   
25%                                 8.000000   
50%                                15.000000   
75%                                23.000000   
max                                31.000000   

       Arrival at Destination - Weekday (Mo = 1)  Distance (KM)   Temperature  \
count                               21201.000000   21201.000000  16835.000000   
mean                                    3.240225       9.506533     23.258889   
std                                     1.567228       5.668963      3.615768   
min                                     1.000000       1.000000     11.200000   
25%                                     2.000000       5.000000     20.600000   
50%                                     3.000000       8.000000     23.500000   
75%                                     5.000000      13.000000     26.000000   
max                                     7.000000      49.000000     32.100000   

       Precipitation in millimeters    Pickup Lat   Pickup Long  \
count                    552.000000  21201.000000  21201.000000   
mean                       7.905797     -1.281470     36.811264   
std                 

In [70]:
# Check for Missing values
df.isna().sum()

Order No                                         0
User Id                                          0
Vehicle Type                                     0
Platform Type                                    0
Personal or Business                             0
Placement - Day of Month                         0
Placement - Weekday (Mo = 1)                     0
Placement - Time                                 0
Confirmation - Day of Month                      0
Confirmation - Weekday (Mo = 1)                  0
Confirmation - Time                              0
Arrival at Pickup - Day of Month                 0
Arrival at Pickup - Weekday (Mo = 1)             0
Arrival at Pickup - Time                         0
Pickup - Day of Month                            0
Pickup - Weekday (Mo = 1)                        0
Pickup - Time                                    0
Arrival at Destination - Day of Month            0
Arrival at Destination - Weekday (Mo = 1)        0
Arrival at Destination - Time  

In [71]:
df.duplicated().sum()

0

### Observations
1. There are two columns with missing values, Temperatures and Precipitation in millimeters

2. The column names need to be Renamed due to long names.

3. There are no duplicates.

4. There are some column names that are are not useful in the data set, e.g Order number and user ID

5. There are some categorical columns that need to be dealt with for example the time columns. These columns need to be converted to date time objects for manipulation

In [72]:
# Dealing with the missing data.
# The precipitation column has so many missing values hence its not valuable to salvage.
# The temperature column can be filled with mean for missing numbers.
# Drop the unecessary columns

df.drop('Precipitation in millimeters', axis= 1 , inplace= True)
df['Temperature'].fillna(df['Temperature'].mean(skipna=True), inplace=True)
df.drop(['Order No','User Id', 'Rider Id'], axis = 1 ,inplace= True)

In [73]:
# Check if missing values have been fixed

df.describe()

Platform Type  Placement - Day of Month  Placement - Weekday (Mo = 1)  \
count   21201.000000              21201.000000                  21201.000000   
mean        2.752182                 15.653696                      3.240083   
std         0.625178                  8.798916                      1.567295   
min         1.000000                  1.000000                      1.000000   
25%         3.000000                  8.000000                      2.000000   
50%         3.000000                 15.000000                      3.000000   
75%         3.000000                 23.000000                      5.000000   
max         4.000000                 31.000000                      7.000000   

       Confirmation - Day of Month  Confirmation - Weekday (Mo = 1)  \
count                 21201.000000                     21201.000000   
mean                     15.653837                         3.240225   
std                       8.798886                         1.567228   
min                       1.000000                         1.000000   
25%                       8.000000                         2.000000   
50%                      15.000000                         3.000000   
75%                      23.000000                         5.000000   
max                      31.000000                         7.000000   

       Arrival at Pickup - Day of Month  Arrival at Pickup - Weekday (Mo = 1)  \
count                      21201.000000                          21201.000000   
mean                          15.653837                              3.240225   
std                            8.798886                              1.567228   
min                            1.000000                              1.000000   
25%                            8.000000                              2.000000   
50%                           15.000000                              3.000000   
75%                           23.000000                              5.000000   
max                           31.000000                              7.000000   

       Pickup - Day of Month  Pickup - Weekday (Mo = 1)  \
count           21201.000000               21201.000000   
mean               15.653837                   3.240225   
std                 8.798886                   1.567228   
min                 1.000000                   1.000000   
25%                 8.000000                   2.000000   
50%                15.000000                   3.000000   
75%                23.000000                   5.000000   
max                31.000000                   7.000000   

       Arrival at Destination - Day of Month  \
count                           21201.000000   
mean                               15.653837   
std                                 8.798886   
min                                 1.000000   
25%                                 8.000000   
50%                                15.000000   
75%                                23.000000   
max                                31.000000   

       Arrival at Destination - Weekday (Mo = 1)  Distance (KM)   Temperature  \
count                               21201.000000   21201.000000  21201.000000   
mean                                    3.240225       9.506533     23.258889   
std                                     1.567228       5.668963      3.222006   
min                                     1.000000       1.000000     11.200000   
25%                                     2.000000       5.000000     21.400000   
50%                                     3.000000       8.000000     23.258889   
75%                                     5.000000      13.000000     25.300000   
max                                     7.000000      49.000000     32.100000   

         Pickup Lat   Pickup Long  Destination Lat  Destination Long  \
count  21201.000000  21201.000000     21201.000000      21201.000000   
mean      -1.281470     36.811264        -1.282581         36.811220   
std  

In [77]:
# View the data set on high level
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21201 entries, 0 to 21200
Data columns (total 25 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Vehicle Type                               21201 non-null  object 
 1   Platform Type                              21201 non-null  int64  
 2   Personal or Business                       21201 non-null  object 
 3   Placement - Day of Month                   21201 non-null  int64  
 4   Placement - Weekday (Mo = 1)               21201 non-null  int64  
 5   Placement - Time                           21201 non-null  object 
 6   Confirmation - Day of Month                21201 non-null  int64  
 7   Confirmation - Weekday (Mo = 1)            21201 non-null  int64  
 8   Confirmation - Time                        21201 non-null  object 
 9   Arrival at Pickup - Day of Month           21201 non-null  int64  
 10  Arrival at Pickup - We

In [79]:
# Check the unique values of vehicle type

df['Vehicle Type'].unique()

array(['Bike'], dtype=object)

In [80]:
# The vehicle type has only one unique value and hence it can be dropped
df.drop('Vehicle Type', axis =1 , inplace= True)

In [81]:
# Check on unique values Personal or Business

df['Personal or Business'].unique()

array(['Business', 'Personal'], dtype=object)

In [82]:
#Chek on distribution of the 2 categories
df['Personal or Business'].value_counts()

Business    17384
Personal     3817
Name: Personal or Business, dtype: int64

In [83]:
# This column is a candidate for Label encoding

from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

df['Personal or Business'] = le.fit_transform(df['Personal or Business'])

In [86]:
# Check the unique categories
df['Personal or Business'].value_counts()

0    17384
1     3817
Name: Personal or Business, dtype: int64

#### Our time feautures are in the form I:M:S, we can convert these to HMS. For example, 2:19:47 PM becomes 141947.

%H is the 24 hour clock, %I is the 12 hour clock and when using the 12 hour clock, %p qualifies if it is AM or PM

from datetime import datetime. 

In [88]:
# Dealing with the time features

def time_converter(time):
  in_time = datetime.strptime(time, "%I:%M:%S %p")
  out_time = datetime.strftime(in_time, "%H%M%S")
  return out_time

time_cols = ['Placement - Time', 'Confirmation - Time', 'Arrival at Pickup - Time', 'Pickup - Time', 'Arrival at Destination - Time']

for col in time_cols:
  df[col] = df[col].apply(time_converter)
  print('Success converting', col)


Success converting Placement - Time
Success converting Confirmation - Time
Success converting Arrival at Pickup - Time
Success converting Pickup - Time
Success converting Arrival at Destination - Time


In [90]:
# Check the resultant Data Frame
df.head()

Platform Type  Personal or Business  Placement - Day of Month  \
0              3                     0                         9   
1              3                     1                        12   
2              3                     0                        30   
3              3                     0                        15   
4              1                     1                        13   

   Placement - Weekday (Mo = 1) Placement - Time  Confirmation - Day of Month  \
0                             5           093546                            9   
1                             5           111616                           12   
2                             2           123925                           30   
3                             5           092534                           15   
4                             1           095518                           13   

   Confirmation - Weekday (Mo = 1) Confirmation - Time  \
0                                5              094010   
1                                5              112321   
2                                2              124244   
3                                5              092605   
4                                1              095618   

   Arrival at Pickup - Day of Month  Arrival at Pickup - Weekday (Mo = 1)  \
0                                 9                                     5   
1                                12                                     5   
2                                30                                     2   
3                                15                                     5   
4                                13                                     1   

   ... Arrival at Destination - Day of Month  \
0  ...                                     9   
1  ...                                    12   
2  ...                                    30   
3  ...                                    15   
4  ...                                    13   

   Arrival at Destination - Weekday (Mo = 1)  Arrival at Destination - Time  \
0                                          5                         103955   
1                                          5                         121722   
2                                          2                         130038   
3                                          5                         100527   
4                                          1                         102537   

  Distance (KM)  Temperature  Pickup Lat Pickup Long  Destination Lat  \
0             4    20.400000   -1.317755   36.830370        -1.300406   
1            16    26.400000   -1.351453   36.899315        -1.295004   
2             3    23.258889   -1.308284   36.843419        -1.300921   
3             9    19.200000   -1.281301   36.832396        -1.257147   
4             9    15.400000   -1.266597   36.792118        -1.295041   

   Destination Long  Time from Pickup to Arrival  
0         36.829741                          745  
1         36.814358                         1993  
2         36.828195                          455  
3         36.795063                         1341  
4         36.809817                         1214  

[5 rows x 24 columns]

In [91]:
# Check the high level DF
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21201 entries, 0 to 21200
Data columns (total 24 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Platform Type                              21201 non-null  int64  
 1   Personal or Business                       21201 non-null  int64  
 2   Placement - Day of Month                   21201 non-null  int64  
 3   Placement - Weekday (Mo = 1)               21201 non-null  int64  
 4   Placement - Time                           21201 non-null  object 
 5   Confirmation - Day of Month                21201 non-null  int64  
 6   Confirmation - Weekday (Mo = 1)            21201 non-null  int64  
 7   Confirmation - Time                        21201 non-null  object 
 8   Arrival at Pickup - Day of Month           21201 non-null  int64  
 9   Arrival at Pickup - Weekday (Mo = 1)       21201 non-null  int64  
 10  Arrival at Pickup - Ti

In [100]:
# Convert the Time columns to Integers

for val in time_cols:
    df[val] = df[val].astype(str).astype(int)

In [101]:
# Check the final Data Frame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21201 entries, 0 to 21200
Data columns (total 24 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Platform Type                              21201 non-null  int64  
 1   Personal or Business                       21201 non-null  int64  
 2   Placement - Day of Month                   21201 non-null  int64  
 3   Placement - Weekday (Mo = 1)               21201 non-null  int64  
 4   Placement - Time                           21201 non-null  int64  
 5   Confirmation - Day of Month                21201 non-null  int64  
 6   Confirmation - Weekday (Mo = 1)            21201 non-null  int64  
 7   Confirmation - Time                        21201 non-null  int64  
 8   Arrival at Pickup - Day of Month           21201 non-null  int64  
 9   Arrival at Pickup - Weekday (Mo = 1)       21201 non-null  int64  
 10  Arrival at Pickup - Ti

### The dataframe is now ready for Machine Learning Modelling.

In [108]:
# Feature selection

X = df.drop('Time from Pickup to Arrival', axis= 1)
y = df['Time from Pickup to Arrival']

In [110]:
# Split the Data Set
    
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2, random_state= 42)

In [111]:
# Import the Random Forest Regressor

from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
import numpy as np

# Initialize the model

base_model = RandomForestRegressor(random_state=42)

# Fit the Data

base_model.fit(X_train,y_train)

# Predict
base_predictor = base_model.predict(X_test)

print('RMSE', np.sqrt(mean_squared_error(base_predictor, y_test)))



RMSE 500.4618540959815


In [114]:
# Model improvement with feature scaling

from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler().fit(X_train)
X_train_normed = scaler.transform(X_train)
X_test_normed = scaler.transform(X_test)

In [115]:
# Fit the normalized model
normalised_model = RandomForestRegressor(random_state=42)
normalised_model.fit(X_train_normed,y_train)
normalised_predictor = normalised_model.predict(X_test_normed)

print('RMSE', np.sqrt(mean_squared_error(normalised_predictor, y_test)))

RMSE 500.96544990245377


In [116]:
# There was a slight  decrease in performance. Lets try modelling with Standadization

from sklearn.preprocessing import StandardScaler

standard = StandardScaler().fit(X_train)
X_train_standardized  = standard.transform(X_train)
X_test_standardized = standard.transform(X_test)

# Fit the model

standardized_model = RandomForestRegressor(random_state=42)
standardized_model.fit(X_train_standardized,y_train)
standadized_predictor = standardized_model.predict(X_test_standardized)

print('RMSE', np.sqrt(mean_squared_error(standadized_predictor, y_test)))


RMSE 500.9073711988233


### The model has improved a bit. Time to try the wrapper methods. 

Step Forward Feature Selection

In [117]:
# Split the data
X = df.drop('Time from Pickup to Arrival', axis=1)
y = df['Time from Pickup to Arrival']

X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2, random_state= 42)

# Normalize the Data

scaler = MinMaxScaler().fit(X_train)
X_train_normed = scaler.transform(X_train)
X_test_normed = scaler.transform(X_test)

In [120]:
 
from mlxtend.feature_selection import SequentialFeatureSelector
    
#Instatiate the model and the Feature selector

sequential_regressor = RandomForestRegressor(random_state=42)

sf_feature_selector = SequentialFeatureSelector(sequential_regressor,
                                               k_features = 4,
                                               forward = True,
                                               verbose = 1,
                                               scoring = 'r2',
                                               cv=4)
sf_feature_selector = sf_feature_selector.fit(X_train_normed,y_train)

[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  23 out of  23 | elapsed:  1.3min finished
Features: 1/4[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  22 out of  22 | elapsed:  1.2min finished
Features: 2/4[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  21 out of  21 | elapsed:  1.3min finished
Features: 3/4[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done  20 out of  20 | elapsed:  1.5min finished
Features: 4/4

In [121]:
sf_feat_cols = list(sf_feature_selector.k_feature_idx_)
print(sf_feat_cols)

[0, 1, 3, 17]


In [122]:
# Modelling with the selected features

regressor_with_selected_forward_features  = RandomForestRegressor(random_state=42)
regressor_with_selected_forward_features.fit(X_train_normed[:,sf_feat_cols],y_train)

#Predictions

selected_forward_features_predictor = regressor_with_selected_forward_features.predict(X_test_normed[:,sf_feat_cols])
print('RMSE with Selective Forward Feature Selection', np.sqrt(mean_squared_error(selected_forward_features_predictor, y_test)))

RMSE with Selective Forward Feature Selection 809.288022341077


#### Observations With 4 features the performance was horrible. Increasing the features will improve the performance. Although this will not be done in this project. Time to try Recursive Feature Elimination


In [125]:
# split data
X = df.drop(['Time from Pickup to Arrival'], axis=1)
y = df['Time from Pickup to Arrival']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.2, random_state=42)

# normalization
norm = MinMaxScaler().fit(X_train) 
X_train_normed = norm.transform(X_train) 
X_test_normed = norm.transform(X_test)

In [126]:
# Instatiating the regressor and RFE object

from sklearn.feature_selection import RFE

Recursive_Regressor = RandomForestRegressor(random_state=42)
Recursive_Regressor = RFE(Recursive_Regressor, n_features_to_select=20, step = 1)

Recursive_Regressor.fit(X_train_normed,y_train)

RFE_predictor = Recursive_Regressor.predict(X_test_normed)

print('RMSE with Recursive Feature Elimination',np.sqrt(mean_squared_error(RFE_predictor,y_test)))

RMSE with Recursive Feature Elimination 500.6372774459313


### Observations: Improvements to 500.63. Time to try out Principal Componet analysis

In [127]:
#prepare the data
X = df.drop(['Time from Pickup to Arrival'], axis=1)
y = df['Time from Pickup to Arrival']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.2, random_state=42)

# normalization
norm = MinMaxScaler().fit(X_train) 
X_train_normed = norm.transform(X_train) 
X_test_normed = norm.transform(X_test)

from sklearn.decomposition import PCA
pca = PCA()
X_train_PCA = pca.fit_transform(X_train_normed)
X_test_PCA = pca.transform(X_test_normed)

# Fit the model

regressor_with_pca = RandomForestRegressor(random_state=42)
regressor_with_pca.fit(X_train_PCA, y_train)

pca_y_pred = regressor_with_pca.predict(X_test_PCA)

# Finally, evaluate our model  
print('RMSE with PCA:', np.sqrt(mean_squared_error(y_test, pca_y_pred)))



RMSE with PCA: 360.65155292646614


#### Massive improvements to 360. This has earned me a cup of coffee!!

### Feature Transformation with LDA

In [128]:
#prepare the data
X = df.drop(['Time from Pickup to Arrival'], axis=1)
y = df['Time from Pickup to Arrival']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.2, random_state=42)

# normalization
norm = MinMaxScaler().fit(X_train) 
X_train_normed = norm.transform(X_train) 
X_test_normed = norm.transform(X_test)

In [130]:
# Instatiate the models
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis


lda = LinearDiscriminantAnalysis(n_components=2)
X_train = lda.fit_transform(X_train_normed, y_train)
X_test = lda.transform(X_test_normed)

classifier = RandomForestRegressor(random_state=42)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print('Random Forest:', np.sqrt(mean_squared_error(y_test, y_pred)))

Random Forest: 358.05387837233695


### Observations: Slight improvements to 358. Time to try out Feature construction

In [131]:
# create a new feature:  speed = distance/time

df['speed'] = df['Distance (KM)'] / (df['Time from Pickup to Arrival'] / 3600)

In [132]:
# Review the df
df.head()

Platform Type  Personal or Business  Placement - Day of Month  \
0              3                     0                         9   
1              3                     1                        12   
2              3                     0                        30   
3              3                     0                        15   
4              1                     1                        13   

   Placement - Weekday (Mo = 1)  Placement - Time  \
0                             5             93546   
1                             5            111616   
2                             2            123925   
3                             5             92534   
4                             1             95518   

   Confirmation - Day of Month  Confirmation - Weekday (Mo = 1)  \
0                            9                                5   
1                           12                                5   
2                           30                                2   
3                           15                                5   
4                           13                                1   

   Confirmation - Time  Arrival at Pickup - Day of Month  \
0                94010                                 9   
1               112321                                12   
2               124244                                30   
3                92605                                15   
4                95618                                13   

   Arrival at Pickup - Weekday (Mo = 1)  ...  \
0                                     5  ...   
1                                     5  ...   
2                                     2  ...   
3                                     5  ...   
4                                     1  ...   

   Arrival at Destination - Weekday (Mo = 1)  Arrival at Destination - Time  \
0                                          5                         103955   
1                                          5                         121722   
2                                          2                         130038   
3                                          5                         100527   
4                                          1                         102537   

   Distance (KM)  Temperature  Pickup Lat  Pickup Long  Destination Lat  \
0              4    20.400000   -1.317755    36.830370        -1.300406   
1             16    26.400000   -1.351453    36.899315        -1.295004   
2              3    23.258889   -1.308284    36.843419        -1.300921   
3              9    19.200000   -1.281301    36.832396        -1.257147   
4              9    15.400000   -1.266597    36.792118        -1.295041   

   Destination Long  Time from Pickup to Arrival      speed  
0         36.829741                          745  19.328859  
1         36.814358                         1993  28.901154  
2         36.828195                          455  23.736264  
3         36.795063                         1341  24.161074  
4         36.809817                         1214  26.688633  

[5 rows x 25 columns]

In [133]:
# Splitting the data
X =  df.drop(['Time from Pickup to Arrival'], axis=1)
y =  df['Time from Pickup to Arrival']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.2, random_state=42)

# normalization
norm = MinMaxScaler().fit(X_train) 
X_train_normed = norm.transform(X_train) 
X_test_normed = norm.transform(X_test)


In [136]:
#Fitting the model
second_base_model = RandomForestRegressor(random_state=42)
second_base_model.fit(X_train_normed,y_train)
new_predictor = second_base_model.predict(X_test_normed)

print('RMSE:', np.sqrt(mean_squared_error(y_test, new_predictor)))


RMSE: 78.20192010411137


### So far this is the Best Performance. Last model is by Step Backward Selection with 20 features

### Conclusion The best results were Obtained after Creating New Features with data Normalization.

### Challenge the Solution

a) Did we have the right question? Yes

b) Did we have the right data? Yes

c) What can be done to improve the solution?

Hyperparameter tuning
Using more features for the Wrapper methods. I needed more computing Power to achieve this.
Handle any outliers